Add travis and deploy script for github pages

MartinHjelmare opened the pull request:
- Add encrypted github deploy key. We need to update this when we've added a new deploy key to the upstream github repo. I've only added it to my fork for now. Where do we want to put this. For now I just put it into the root. Maybe put it into a hidden directory `.travis`?
- Add deploy script. We can change the directory for scripts if we want.

Here are my test results:
https://travis-ci.org/MartinHjelmare/ImJoy/builds/460874796
https://github.com/MartinHjelmare/ImJoy/tree/gh-pages

**TODO**
- [x] Add a new public deploy key to upstream github repo settings.
  https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/#generating-a-new-ssh-key
  https://developer.github.com/v3/guides/managing-deploy-keys/#deploy-keys
- [x] Update the deploy key when we've added a new deploy key to the upstream github repo.
  https://docs.travis-ci.com/user/encrypting-files/
- [x] Activate travis for the upstream repo.
- [x] 1. move a folder `/docs/docs` to `/web/public/docs`, because it contains markdown source code, and it should kept on the master branch.
- [x] 2. copy the following files from `/docs` to `/web/public`: `.nojekyll`, `404.html`, `CNAME`, `manifest.json`.
- [x] 3. change the following line in `/web/package.json`:

     `"build": "rimraf ../docs/{css,js,static,index.html} && vue-cli-service build --no-clean",`

     to:

     `"build": "vue-cli-service build",`
- [ ] ~~Remove `/docs`.~~

oeway commented:
Hey, thanks for working on this. It seems you already get it work, cool.

I think put the `.travis.yml` is fine for now, however, considering there maybe more files to come (including the deploy script, the `.enc` file, and potentially more in the future), how about make a `utils` folder for all these utility stuff? Would that work for Travis?

Another thing is, the deployment, for now when we run `npm run build`, it will build the website and generate static files (html/css/etc.) copy them to `/docs`, and in Github Pages I set it to serve from `/docs` of master branch. Maybe related to that, each time I push a new build, Github pages do not serve the new version immediately, I had to do another empty commit and push again (that why you see many `rebuild pages` commits which makes the history unreadable). Since we can use Travis to build the pages, and we can build it to a new branch, how about expand those files in `/docs` of current `master` branch to `gh-pages` branch? And then we set Github pages to serve the entire `gh-pages` branch. This would make `gh-pages` contains all generated files and `master` contains only source code.

If we do that, we may also need to make an additional changes:
1. move a folder `/docs/docs` to `/web/public/docs`, because it contains markdown source code, and it should kept on the master branch.
2. copy the following files from `/docs` to `/web/public`: `.nojekyll`, `404.html`, `CNAME`, `manifest.json`.
3. change the following line in `/web/package.json`:

     `"build": "rimraf ../docs/{css,js,static,index.html} && vue-cli-service build --no-clean",`

     to:

     `"build": "vue-cli-service build",`

MartinHjelmare commented on scripts/deploy:
```diff
@@ -0,0 +1,55 @@
+
+# Clone the existing gh-pages for this repo into gh-pages/
+# Create a new empty branch if gh-pages doesn't exist yet (should only happen on first deploy).
+git clone $REPO gh-pages
```
Here we clone the repo into `gh-pages` directory.

MartinHjelmare commented on scripts/deploy:
```diff
@@ -0,0 +1,55 @@
+# Clone the existing gh-pages for this repo into gh-pages/
+# Create a new empty branch if gh-pages doesn't exist yet (should only happen on first deploy).
+git clone $REPO gh-pages
+cd gh-pages
```
We move into that dir.

MartinHjelmare commented on scripts/deploy:
```diff
@@ -0,0 +1,55 @@
+# Create a new empty branch if gh-pages doesn't exist yet (should only happen on first deploy).
+git clone $REPO gh-pages
+cd gh-pages
+git checkout $TARGET_BRANCH || git checkout --orphan $TARGET_BRANCH
```
Checkout the `gh-branch`.

MartinHjelmare commented on scripts/deploy:
```diff
@@ -0,0 +1,55 @@
+# UPDATE THIS!
+# cp -Rf $HOME/dist/* .
+# Copy dirs and files and that we want to update.
+cp -Rf ../docs/* ./docs/
```
And copy all files in `master` root `/docs` to `gh-pages/docs` where `gh-pages/` is the root of the deploy `gh-pages` branch (repo).

MartinHjelmare replied:
We're only copying `master/docs` to `gh-pages/docs` but we are copying to a directory where we have cloned the repo and checked out a new orphan branch `gh-pages`, so that will contain all the contents of master including `/web`.

We could create the `gh-pages` branch manually first and only commit the docs files outside of the deploy script. Then when the script would run, it would just update the docs files.

I'll change it to how you suggest.

oeway replied:
Great, thanks. I think one reason is that Github pages seems only support serving from a `/docs` folder in master branch, then the other choice is serve the entire `gh-pages` branch ([ref](https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/)). That would require to remove the `/docs` folder from `gh-pages` and take out all the files to the root of `gh-pages`.

MartinHjelmare replied:
Good! I'll change to that dir structure.

MartinHjelmare commented on scripts/deploy:
```diff
@@ -0,0 +1,55 @@
+git commit -m "Deploy to GitHub Pages $TARGET_BRANCH branch: ${SHA}"
+
+# Now that we're all set up, we can push.
+git push $SSH_REPO $TARGET_BRANCH
```
We add everything, commit and push to `gh-pages'.

MartinHjelmare commented:
@oeway If we want the feature to be functional directly when it's merged you as owner needs to add the a deploy key as described above.

Should we do this in a separate PR first or do you want to push to this branch? Either works.

oeway commented:
Let's do it now, would be great to have it ASAP.

MartinHjelmare commented:
> If we do that, we may also need to make an additional changes:
>
> 1. move a folder /docs/docs to /web/public/docs, because it contains markdown source code, and it should kept on the master branch.
>
> 2. copy the following files from /docs to /web/public: .nojekyll, 404.html, CNAME, manifest.json.
>
> 3. change the following line in /web/package.json:
>
>     "build": "rimraf ../docs/{css,js,static,index.html} && vue-cli-service build --no-clean",
>
>     to:
>
>     "build": "vue-cli-service build",

It looks like we don't need to do step 1 - 3 from your comment, correct? The master tree will not change structure after this PR.

MartinHjelmare commented:
Or is that's why it's failing now?
https://travis-ci.org/oeway/ImJoy/builds/461791164

oeway commented:
We will need to do that, because the current setup in the master tree tries to keep some markdown files `/docs/docs` during each build. Now if we move `/docs/docs` to `/web/public/docs`, we can just remove the entire `/docs` folder in the master branch.

I am not sure if that build error related to that, can you try to increase the heap size? [ref](https://stackoverflow.com/questions/38558989/node-js-heap-out-of-memory)

MartinHjelmare commented:
So I've removed followed step 1 - 3 as instructed and also removed the `/docs` directory. I'm a little hesitant about including the last part in this PR, since the diff is so big after all the removed files.

What do you think?

oeway commented:
I think that's fine, the `/docs` are mostly generated files, please go ahead and merge it.
